components/admin/lessons/AdminLessonInputs.tsx

```diff
@@ -0,0 +1,114 @@
+import React, { useState } from 'react'
+import { useAddModuleMutation } from '../../../graphql'
+import { formChange } from '../../../helpers/formChange'
+import { FormCard, MD_INPUT, Option } from '../../FormCard'
+import { AlertFillIcon, CheckCircleIcon } from '@primer/octicons-react'
+import styles from '../../../scss/adminLessonInputs.module.scss'
+import { Spinner } from 'react-bootstrap'
+import { GraphQLError } from 'graphql-request/dist/types'
+
+type Props = {
+  lesson: { title: string; id: number }
+}
+
+enum Error {
+  InvalidData = 'missing module name or description'
+}
+
+type FormOptions = (Option & { value?: string })[]
```

Comment:
Why is `& { value?: string }` needed? Seems there is already a value field in the Option type definition
https://github.com/garageScript/c0d3-app/blob/master/components/FormCard.tsx#L20

Reply:
Ah, I see. Can you try passing a type variable to the useState call for formOptions? And see if that eliminates the need to do `& { value?: string }`

```
useState<Option[]>(
```

I'm thinking that in the future, there is potential for additional fields added to the form. And one of these additional fields might be a dropdown menu, so it might not be a good idea to restrict value to just `string`

Reply:
`useState<Option[]>(values)` is similar to how it originally was _as_ TypeScript inferred the type as `Option[]`. Now we'll have to add type casting to prevent it from throwing `Type 'string | number | Item[]' is not assignable to type 'string'`.

Reply:
Hmm I see.. I'll try to look more into this today to see if I can find a solution

Reply:
any update on this?

Reply:
A few options:

### Typescript 

Pros:
* With this method, we correctly define the type in the values array.
* Doing things this way will enable us to add a dropdown menu option in the future.

Cons:
* If this array grows large, it can be a readability concern to have to manually add the types for each element in the array. I am predicting that this array will not grow too large though, since this form is only for a specific use case. 
```
type textField = Omit<Option, 'value'> & {value: string}
const values: [textField, textField] = [
  {
    title: 'Module Name',
    value: '' // notice here that we define initial value as empty string for `value` field, instead of giving it a fallback value of empty string when passing it as a variable to the graphql mutation
  },
  {
    title: 'Description',
    type: 'MD_INPUT',
    value: ''
  }
]
```

### Using typeof operator

Pros:
* We don't have to make any new type definitions

Cons:
* We are doing the type checking ourselves, which seems like an anti-pattern because are doing the job that typescript should already do
* Every time a new field gets added to this form, we will have to do typechecking for that new field as well. If a dropdown menu gets added as a field, we would have to implement logic to check for type, and pull the right value correspondingly.

### Using String function

Pros:
* We don't have to make any new type definitions

Cons: 
* If the field is already a string, it seems redundant to have to stringify the string.
* If a dropdown menu gets added, a developer could fall into the problem of using String function on the dropdown menu value(which is an array). This can potentially cause a future issue of not properly grabbing value from the dropdown menu array.

### My opinion
I like the typescript solution the best. This solution allows us to still add a dropdown menu field to this form in the future, and we won't be skipping any type checking either, thus there will be no decrease in type safety. Since these form fields do not have dropdown menus yet, type casting field values to strings is actually not a huge safety concern. The potential for problems arises if we add a dropdown menu in the future. If we do type casting for those dropdown menus(they have an array of Item type ass their type), we can run into issues since type checking will not be strict. This is why we should avoid setting the precedent of using type casting in this file, and should only use it if it is the most appropriate choice to use

Reply:
Nice solution. It even solves another issue I had. I should learn more about TypeScript utilities.